### DISCUSSION TITLE
First Opening is Slow

### DISCUSSION THREAD

**Discussion by Tahinli** [ghostty-org/ghostty#2978] [Q&A] [answered]
Hi, thanks for Ghostty at the first step.

Opening Ghostty takes 2-3x times more then kitty, alacritty and wezterm.

My specs are: 

10750H
1660ti driver version: 565.77 
Fedora 41 & Hyprland

startup logs are here:

```text
info: ghostty version=0.1.0-dev+0000000
info: ghostty build optimize=ReleaseFast
info: runtime=apprt.Runtime.gtk
info: font_backend=font.main.Backend.fontconfig_freetype
info: dependency harfbuzz=8.4.0
info: dependency fontconfig=21402
info: renderer=renderer.OpenGL
info: libxev backend=main.Backend.io_uring
info(os): setlocale from env result=en_GB.UTF-8
info(gtk): GTK version build=4.16.5 runtime=4.16.5
info: reading configuration file path=/home/tahinli/.config/ghostty/config
info(config): config has 'keybind = clear', all keybinds cleared
info(config): default shell source=env value=/bin/zsh
EGL API version 1.5 found
 - Vendor: NVIDIA
 - Version: 1.5
 - Device: /dev/dri/renderD128, 226 128
 - Client APIs: OpenGL_ES OpenGL
 - Extensions:
	EGL_ANDROID_native_fence_sync
	EGL_EXT_buffer_age
	EGL_EXT_client_sync
	EGL_EXT_create_context_robustness
	EGL_EXT_image_dma_buf_import
	EGL_EXT_image_dma_buf_import_modifiers
	EGL_EXT_output_base
	EGL_EXT_output_drm
	EGL_EXT_present_opaque
	EGL_EXT_protected_content
	EGL_EXT_stream_acquire_mode
	EGL_EXT_stream_consumer_egloutput
	EGL_EXT_sync_reuse
	EGL_IMG_context_priority
	EGL_KHR_config_attribs
	EGL_KHR_context_flush_control
	EGL_KHR_create_context
	EGL_KHR_create_context_no_error
	EGL_KHR_fence_sync
	EGL_KHR_get_all_proc_addresses
	EGL_KHR_gl_colorspace
	EGL_KHR_gl_renderbuffer_image
	EGL_KHR_gl_texture_2D_image
	EGL_KHR_gl_texture_3D_image
	EGL_KHR_gl_texture_cubemap_image
	EGL_KHR_image
	EGL_KHR_image_base
	EGL_KHR_no_config_context
	EGL_KHR_partial_update
	EGL_KHR_reusable_sync
	EGL_KHR_stream
	EGL_KHR_stream_attrib
	EGL_KHR_stream_consumer_gltexture
	EGL_KHR_stream_cross_process_fd
	EGL_KHR_stream_fifo
	EGL_KHR_stream_producer_eglsurface
	EGL_KHR_surfaceless_context
	EGL_KHR_swap_buffers_with_damage
	EGL_KHR_wait_sync
	EGL_MESA_image_dma_buf_export
	EGL_NV_nvrm_fence_sync
	EGL_NV_output_drm_flip_event
	EGL_NV_quadruple_buffer
	EGL_NV_robustness_video_memory_purge
	EGL_NV_stream_attrib
	EGL_NV_stream_consumer_eglimage
	EGL_NV_stream_consumer_gltexture_yuv
	EGL_NV_stream_cross_display
	EGL_NV_stream_cross_object
	EGL_NV_stream_cross_process
	EGL_NV_stream_cross_system
	EGL_NV_stream_dma
	EGL_NV_stream_fifo_next
	EGL_NV_stream_fifo_synchronous
	EGL_NV_stream_flush
	EGL_NV_stream_metadata
	EGL_NV_stream_origin
	EGL_NV_stream_remote
	EGL_NV_stream_reset
	EGL_NV_stream_socket
	EGL_NV_stream_socket_inet
	EGL_NV_stream_socket_unix
	EGL_NV_stream_sync
	EGL_NV_system_time
	EGL_NV_triple_buffer
	EGL_WL_bind_wayland_display
	EGL_WL_wayland_eglstream
 - Selected fbconfig: R8G8B8A8, depth 0, stencil 0
          high depth: none
Creating EGL context version 3.2 (debug:no, forward:no, legacy:no, es:no)
Using OpenGL backend EGL
Creating EGL context version 3.2 (debug:no, forward:no, legacy:no, es:no)
OpenGL version: 4.6 (core)
GLSL version: 4.60 NVIDIA
Max texture size: 32768
Enabled features (use GDK_GL_DISABLE env var to disable):
    debug: ✓
    unpack-subimage: ✓
    half-float: ✓
    sync: ✓
    base-instance: ✓
    buffer-storage: ✓
OpenGL version: 4.6 (core)
GLSL version: 4.60 NVIDIA
Max texture size: 32768
Enabled features (use GDK_GL_DISABLE env var to disable):
    debug: ✓
    unpack-subimage: ✓
    half-float: ✓
    sync: ✓
    base-instance: ✓
    buffer-storage: ✓
Creating EGL context version 3.3 (debug:no, forward:no, legacy:no, es:no)
OpenGL version: 4.6 (core)
GLSL version: 4.60 NVIDIA
Max texture size: 32768
Enabled features (use GDK_GL_DISABLE env var to disable):
    debug: ✓
    unpack-subimage: ✓
    half-float: ✓
    sync: ✓
    base-instance: ✓
    buffer-storage: ✓
info(grid): loaded OpenGL 4.6
info(io_exec): found Ghostty resources dir: /home/tahinli/.local/bin/share/ghostty
info(io_exec): shell integration automatically injected shell=termio.shell_integration.Shell.zsh
warning(gtk): unimplemented action=apprt.action.Action.Key.cell_size
warning(gtk): unimplemented action=apprt.action.Action.Key.size_limit
info(io_exec): started subcommand path=/bin/sh pid=100917
info(io_exec): subcommand cgroup=-
Creating EGL context version 3.2 (debug:no, forward:no, legacy:no, es:no)
OpenGL version: 4.6 (core)
GLSL version: 4.60 NVIDIA
Max texture size: 32768
Enabled features (use GDK_GL_DISABLE env var to disable):
    debug: ✓
    unpack-subimage: ✓
    half-float: ✓
    sync: ✓
    base-instance: ✓
    buffer-storage: ✓
Using fractional scale 1 for EGL window (960 523 => 960 523)
Using fractional scale 1 for EGL window (931 494 => 931 494)
info(grid): reallocating GPU buffer old=0 new=8
info(osc): unknown semantic prompts option: cl
warning(osc): invalid OSC command: 1337;SetUserVar=WEZTERM_PROG=
warning(osc): invalid OSC command: 1337;SetUserVar=WEZTERM_USER=dGFoaW5saQ==
warning(osc): invalid OSC command: 1337;SetUserVar=WEZTERM_IN_TMUX=MA==
warning(osc): invalid OSC command: 1337;SetUserVar=WEZTERM_HOST=dGZlZG9yYQ==
info(stream): OSC 1 (change icon) received and ignored icon=~
warning(stream): unimplemented OSC command: change_window_icon
warning(osc): invalid OSC command: 133;P;k=i
info(grid): reallocating GPU buffer old=8 new=38

(process:100894): Gdk-CRITICAL **: 01:00:59.439: gdk_event_get_modifier_state: assertion 'GDK_IS_EVENT (event)' failed
```

**Comment by mitchellh** [comment]
I converted to Q&A until more details are grabbed. 

I would expect our _first window_ startup time to be significantly slower than WezTerm or Kitty, but can't quantify that. We're using GTK, which is an extremely heavyweight GUI toolkit and initialization takes a lot of time. 

I haven't compared to WezTerm or Kitty, but I know that _simply initializing GTK_ (i.e. zero Ghostty code) is slower than `foot`'s _entire_ first paint startup time, for example. Nothing we can do about that.

So I'm just making this a Q&A to figure out if this is just GTK being GTK or if there is more to it.

As an aside our N=2+ window startup time should be extremely fast, competitive with WezTerm and Kitty, since GTK only initializes for the first window.

**Reply by gptlang** [reply]
> As an aside our N=2+ window startup time should be extremely fast, competitive with WezTerm and Kitty, since GTK only initializes for the first window.

Not sure if it's just on my end but N=2+ startup time is similarly slow (around 300ms). Kitty and wezterm gets 150ms. Foot ~70ms.

It just feels quite painful. Is there any way to disable features such as font ligatures to make it start up faster?

```
info: ghostty version=1.0.1-main+a8e5eef
info: ghostty build optimize=ReleaseFast
info: runtime=apprt.Runtime.gtk
info: font_backend=font.main.Backend.fontconfig_freetype
info: dependency harfbuzz=8.4.0
info: dependency fontconfig=21402
info: renderer=renderer.OpenGL
info: libxev backend=main.Backend.io_uring
info(os): setlocale from env result=en_US.UTF-8
info(gtk): GTK version build=4.16.5 runtime=4.16.5
info: reading configuration file path=/home/acheong/.config/ghostty/config
info(config): default shell source=env value=/usr/bin/fish
info(gtk): libadwaita version build=1.6.2 runtime=1.6.2
EGL API version 1.5 found
 - Vendor: Mesa Project
 - Version: 1.5
 - Device: /dev/dri/renderD128, 226 128
 - Client APIs: OpenGL OpenGL_ES
 - Extensions:
        EGL_ANDROID_blob_cache
        EGL_ANDROID_native_fence_sync
        EGL_EXT_buffer_age
        EGL_EXT_config_select_group
        EGL_EXT_create_context_robustness
        EGL_EXT_image_dma_buf_import
        EGL_EXT_image_dma_buf_import_modifiers
        EGL_EXT_pixel_format_float
        EGL_EXT_present_opaque
        EGL_EXT_protected_content
        EGL_EXT_query_reset_notification_strategy
        EGL_EXT_surface_compression
        EGL_EXT_swap_buffers_with_damage
        EGL_IMG_context_priority
        EGL_KHR_cl_event2
        EGL_KHR_config_attribs
        EGL_KHR_context_flush_control
        EGL_KHR_create_context
        EGL_KHR_create_context_no_error
        EGL_KHR_fence_sync
        EGL_KHR_get_all_proc_addresses
        EGL_KHR_gl_colorspace
        EGL_KHR_gl_renderbuffer_image
        EGL_KHR_gl_texture_2D_image
        EGL_KHR_gl_texture_3D_image
        EGL_KHR_gl_texture_cubemap_image
        EGL_KHR_image_base
        EGL_KHR_no_config_context
        EGL_KHR_reusable_sync
        EGL_KHR_surfaceless_context
        EGL_KHR_swap_buffers_with_damage
        EGL_KHR_wait_sync
        EGL_MESA_configless_context
        EGL_MESA_drm_image
        EGL_MESA_gl_interop
        EGL_MESA_image_dma_buf_export
        EGL_MESA_query_driver
        EGL_MESA_x11_native_visual_id
        EGL_WL_bind_wayland_display
        EGL_WL_create_wayland_buffer_from_image
 - Selected fbconfig: R8G8B8A8, depth 0, stencil 0
          high depth: R16G16B16A16 float, depth 0, stencil 0
Creating EGL context version 3.2 (debug:no, forward:no, legacy:no, es:no)
Using OpenGL backend EGL
Creating EGL context version 3.2 (debug:no, forward:no, legacy:no, es:no)
OpenGL version: 4.6 (core)
GLSL version: 4.60
Max texture size: 16384
Enabled features (use GDK_GL_DISABLE env var to disable):
    debug: ✓
    unpack-subimage: ✓
    half-float: ✓
    sync: ✓
    base-instance: ✓
    buffer-storage: ✓
OpenGL version: 4.6 (core)
GLSL version: 4.60
Max texture size: 16384
Enabled features (use GDK_GL_DISABLE env var to disable):
    debug: ✓
    unpack-subimage: ✓
    half-float: ✓
    sync: ✓
    base-instance: ✓
    buffer-storage: ✓
Creating EGL context version 3.3 (debug:no, forward:no, legacy:no, es:no)
OpenGL version: 4.6 (core)
GLSL version: 4.60
Max texture size: 16384
Enabled features (use GDK_GL_DISABLE env var to disable):
    debug: ✓
    unpack-subimage: ✓
    half-float: ✓
    sync: ✓
    base-instance: ✓
    buffer-storage: ✓
info(grid): loaded OpenGL 4.6
info(font_shared_grid_set): font regular: JetBrainsMono NFM Regular
info(font_shared_grid_set): font bold: JetBrainsMono NFM Bold
info(font_shared_grid_set): font italic: JetBrainsMono NFM Italic
info(font_shared_grid_set): font bold_italic: JetBrainsMono NFM Bold Italic
warning(io_exec): shell could not be detected, no automatic shell integration will be injected
warning(gtk): unimplemented action=apprt.action.Action.Key.cell_size
warning(gtk): unimplemented action=apprt.action.Action.Key.size_limit
info(io_exec): started subcommand path=/bin/sh pid=439309
info(io_exec): subcommand cgroup=-
info(font_shared_grid_set): font regular: JetBrainsMono NFM Regular
info(font_shared_grid_set): font bold: JetBrainsMono NFM Bold
info(font_shared_grid_set): font italic: JetBrainsMono NFM Italic
info(font_shared_grid_set): font bold_italic: JetBrainsMono NFM Bold Italic
warning(gtk): unimplemented action=apprt.action.Action.Key.cell_size
Creating EGL context version 3.2 (debug:no, forward:no, legacy:no, es:no)
OpenGL version: 4.6 (core)
GLSL version: 4.60
Max texture size: 16384
Enabled features (use GDK_GL_DISABLE env var to disable):
    debug: ✓
    unpack-subimage: ✓
    half-float: ✓
    sync: ✓
    base-instance: ✓
    buffer-storage: ✓
Using fractional scale 1 for EGL window (564 712 => 564 712)
Using fractional scale 2 for EGL window (564 712 => 1128 1424)
Using fractional scale 2 for EGL window (561 710 => 1122 1420)
info(font_shared_grid_set): font regular: JetBrainsMono NFM Regular
info(font_shared_grid_set): font bold: JetBrainsMono NFM Bold
info(font_shared_grid_set): font italic: JetBrainsMono NFM Italic
info(font_shared_grid_set): font bold_italic: JetBrainsMono NFM Bold Italic
warning(gtk): unimplemented action=apprt.action.Action.Key.cell_size
info(grid): reallocating GPU buffer old=0 new=20
Destroying EGL context
Destroying EGL context
info(io_exec): read thread got quit signal
info(surface): surface closed addr=3d36dab0
info(sentry): sentry envelope does not contain crash, discarding

________________________________________________________
Executed in    3.62 secs      fish           external
   usr time  344.47 millis  652.00 micros  343.81 millis
   sys time  118.78 millis    0.00 micros  118.78 millis
```

There's also a weird issue where my cursor is 2x the size on the ghostty window. Might be a GTK bug though

**Reply by mmirus** [reply]
I am also seeing the slow startup time in every case, not just on the first window. Running ghostty with `--gtk-single-instance=true` results in the expected behavior: slow on first launch, then fast.

This implies to me that ghostty isn't correctly detecting when it should automatically use single-instance mode. It may be that my environment is not typical enough to be detectable yet: I'm running sway on Arch Linux, and launching ghostty from rofi (which uses the .desktop file). Perhaps if I were in Gnome, this would be working correctly!

Happy to follow up here with additional information if helpful--let me know. And thanks for ghostty!

**Reply by mitchellh** [reply]
Here is the function we use to detect if we should use single instance by default: https://github.com/ghostty-org/ghostty/blob/6cbd69da7839260508466f9dfb2bc0c0fbb43991/src/os/desktop.zig#L10-L19

If you can determine why that doesn't work for you I'd love to improve it.

**Reply by mmirus** [reply]
Thanks, Mitchell. I think `GIO_LAUNCHED_DESKTOP_FILE_PID` is not defined in my ghostty instance. I checked this by launching ghostty and running `env | grep GIO_LAUNCHED_DESKTOP_FILE_PID`. Am I correct that we would expect it to show up there if my system set it when launching ghostty from the `.desktop` file?

**Reply by mitchellh** [reply]
My understanding is at the very least Gnome Desktop sets that for desktop-launched software. What DE are you using? We of course support any desktop environment but may need to add additional detection for others. Do you see any way in your env to detect this?

**Reply by mmirus** [reply]
I'm using [sway](https://github.com/swaywm/sway/), which is a more minimal tiling window manager--not a full desktop environment. For launching applications from `.desktop` files, I'm using a program named [rofi](https://github.com/lbonn/rofi).

For now, I'm working around this by modifying the `.desktop` file to include `--gtk-single-instance=true`, which is working well, but if I can help figure out a more plug-and-play solution, I'm happy to do so.

I'll have to look into if that combination supports setting `GIO_LAUNCHED_DESKTOP_FILE_PID`, or if there are other good options for detecting ghostty desktop launches in this context. I'm a little out of my wheelhouse, though.

**Reply by jcollie** [reply]
I've played briefly with `river` and I didn't see anything obvious that was set when launching applications that could be used to detect a launch. Other DEs might be different. But I would definitely expect that any DE that _does not_ use the `.desktop` file to launch won't set `GIO_LAUNCHED_DESKTOP_FILE_PID`.

Also, `ghostty --gtk-single-instance=true --quit-after-last-window-closed=false --initial-window=false` will run Ghostty in the background without opening a window. Arrange to run that during login, and make sure to launch Ghostty with `--gtk-single-instance=true` (perhaps by setting it in the config) and you should get fast creation of new windows (even the first one) because the initial load time is "absorbed" by the login.

**Reply by jcollie** [reply]
> I'm using [sway](https://github.com/swaywm/sway/), which is a more minimal tiling window manager--not a full desktop environment. For launching applications from `.desktop` files, I'm using a program named [rofi](https://github.com/lbonn/rofi).

Get a dump of all environment variables by running `set` in the shell. Basically we need to find an environment variable that `rofi` sets that won't appear elsewhere. If such an environment variable exists we can use it to detect launches.

**Reply by mmirus** [reply]
Unfortunately, I don't see anything in there that's obviously related to rofi. If you want me to share the whole list for your review, I can take some time to sanitize it of secrets, etc., and then post it.

Does it make any sense at all to consider flipping the logic: instead of checking if ghostty was launched from a desktop environment, detect if ghostty was *not*? Basically, if that comes down to the `ghostty -e "some command"` example mentioned in [the docs](https://ghostty.org/docs/help/gtk-single-instance#desktop-environment-detection), ghostty would be able to ensure that when run with `-e` it can tell it wasn't run from a desktop environment. No idea what other cases would need to be covered here; it might end up being equally difficult, or worse.

PS - thanks for the suggestion about running ghostty in the background during login.

**Reply by trygveaa** [reply]
Can't the command in the `.desktop` file just specify `--gtk-single-instance=true` explicitly?

**Reply by mmirus** [reply]
Yep, that's exactly what I am doing currently as a workaround (and also using that flag in my sway keybinding to launch a terminal).

**Reply by mmirus** [reply]
Recapping some of the options:

1. Flip the logic to default to "ghostty was luanched from a desktop," and detect if it was *not* (e.g., `ghostty -e`), instead of the opposite.
2. Keep the logic as-is, but find additional ways to detect when ghostty was launched from a desktop.
3. Modify the `.desktop` file shipped with ghostty to explicitly include `--gtk-single-instance=true`, per @trygveaa [above](https://github.com/ghostty-org/ghostty/discussions/2978#discussioncomment-11687668).*

Quite possibly, the third option (being explicit in the `.desktop` file) would be good to do in combination with any sort of automatic detection, if there aren't any obvious downsides.

*Note that this wouldn't solve for methods of launching ghostty that don't use the `.desktop` file. For example, it's common to have a global hotkey for launching a terminal. Users can include `--gtk-single-instance=true` explicitly when defining their keybinding, but an ideal solution wouldn't require users to know to do that.

**Reply by mitchellh** [reply]
These are all good options. I just want to note I'm reading these issues but working on catching up on some other things before doing anything. 😄

**Reply by mmirus** [reply]
Thanks, Mitchell. Super impressed with your responsiveness; I'm sure there's been a ton of chatter here following the launch and you have much bigger fish to fry than this topic.

**Reply by jcollie** [reply]
See #2442 for a discussion that uses DBus activation & systemd user units to set things up for much faster operation. I even recently discovered how to set up a global hotkey under Gnome that will open a new Ghostty window, even if Ghostty is not running.

**Reply by iamabdulaziz100** [reply]
"So I'm just making this a Q&A to figure out if this is just GTK being GTK or if there is more to it."

I tried other gtk based option like gnome-console, gnome-terminal and tilix all of them launch faster that gostty

**Comment by choucavalier** [comment]
I just started using ghostty, switching from foot (which does not support font ligatures).

This is the first thing I noticed. Foot opens instantly. Ghostty takes a lot of time to open (what feels like 1 second compared to foot).

This is a bit disappointing because startup time is really important for a terminal: we're spending our days opening terminals to run commands, at least that's my workflow.

Right now I feel like I have to wait on Ghostty to open and I feel like the software is slowing me down.

**Reply by rhodes-b** [reply]
This is the expected behavior as Mitchell has stated the workflow of ghostty is to keep it open for long periods of time 
https://ghostty.org/docs/help/gtk-single-instance
above will significantly help after that first window. If you say the first window is still too slow you can run it without the GUI and keep it alive in the background with a command like this
`ghostty --initial-window=false --quit-after-last-window-closed=false`

**Comment by trygveaa** [comment]
I don't see any difference in startup time when using `--gtk-single-instance=true`. I first started one instance of ghostty with `ghostty --gtk-single-instance=true`. Then to measure startup time I'm running `time ghostty --gtk-single-instance=true -e sh -c exit`. This takes around 0.52 seconds on my machine (when using the powersave scaling governor) regardless of if I specify the `gtk-single-instance` option or not. This is on Arch Linux with the package from Arch's extra repo.

For reference, here's a comparison with some other terminal emulators:

| Command                                                 | Time          |
| ------------------------------------------------------- | ------------- |
| `time ghostty --gtk-single-instance=true -e sh -c exit` | 0.52 s        |
| `time ghostty -e sh -c exit`                            | 0.52 s        |
| `time kitty -e sh -c exit`                              | 0.46 s        |
| `time kitty -1 -e sh -c exit`                           | 0.07 s        |
| `time wezterm start --always-new-process -e sh -c exit` | 0.13 s        |
| `time wezterm start -e sh -c exit`                      | 0.08 s        |
| `time foot -e sh -c exit`                               | 0.09 s        |
| `time alacritty -e sh -c exit`                          | 0.21 s        |
| `time xterm -e sh -c exit`                              | 0.06 s        |

The faster time for kitty and wezterm is when an existing process is running and it opens in that. As far as I can see, foot, alacritty and xterm doesn't have support for that.

**Reply by SuperDuperDeou** [reply]
Can replicate on my machine. After running `ghostty --gtk-single-instance=true --quit-after-last-window-closed=false --initial-window=false`, `ghostty --gtk-single-instance=true -e sh -c exit` takes 0.40~0.45s, the same as simply `ghostty -e sh -c exit`

Additionally, `ghostty --gtk-single-instance=true` exits in 0.08~0.10s while another ghostty --gtk-single-instance is running.

Edit:
`ghostty --gtk-single-instance=true -e sh` does not attach to the existing Ghostty instance. Idk if it's intended behavior or not.

**Reply by trygveaa** [reply]
Ah! Yeah, I see now that if I just run `ghostty --gtk-single-instance=true` it's fast. I suppose I didn't test that, only tested from the launcher where it seemed slow (I'm using rofi on sway, so same issue as mentioned elsewhere in this thread) and with `-e` from another terminal.

I also realize that the times above for kitty and wezterm with an existing process aren't really that useful or comparable to the other times, as the processes exists immediately, so they don't measure how long it takes for the window and shell to be ready.

**Reply by SuperDuperDeou** [reply]
There is a lot to read, so I haven't gone through all the documentation yet, but, I finally stumbled upon something about -e.
```
       --initial-command
              This is the same as “command”, but only applies to the first terminal surface created when Ghostty starts.  Subsequent terminal surfaces will use the command con‐
              figuration.

              After  the first terminal surface is created (or closed), there is no way to run this initial command again automatically.  As such, setting this at runtime works
              but will only affect the next terminal surface if it is the first one ever created.

              If you’re using the ghostty CLI there is also a shortcut to set this with arguments directly: you can use the -e flag.  For example: ghostty -e fish --with --cus‐
              tom --args.  The -e flag automatically forces some other behaviors as well:

              • gtk-single-instance=false - This ensures that a new instance is launched and the CLI args are respected.

              • quit-after-last-window-closed=true - This ensures that  the  Ghostty  process  will  exit  when  the  command  exits.   Additionally,  the  quit-after-last-win‐
                dow-closed-delay is unset.

              • shell-integration=detect  (if  not none) - This prevents forcibly injecting any configured shell integration into the command’s environment.  With -e its highly
                unlikely that you’re executing a shell and forced shell integration is likely to cause problems (i.e. by wrapping your command in a  shell,  setting  env  vars,
                etc.).  This is a safety measure to prevent unexpected behavior.  If you want shell integration with a -e-executed command, you must either name your binary ap‐
                popriately or source the shell integration script manually.

```

So, -e overriding --gtk-single-instance is intended behavior. I have tried `ghostty --gtk-single-instance=true --command="sh -c exit"` but that just runs a standard $SHELL session, while `ghostty --command="sh -c exit"` behaves as expected.

**Reply by rhodes-b** [reply]
> Ah! Yeah, I see now that if I just run `ghostty --gtk-single-instance=true` it's fast. I suppose I didn't test that, only tested from the launcher where it seemed slow (I'm using rofi on sway, so same issue as mentioned elsewhere in this thread) and with `-e` from another terminal.
> 
> I also realize that the times above for kitty and wezterm with an existing process aren't really that useful or comparable to the other times, as the processes exists immediately, so they don't measure how long it takes for the window and shell to be ready.

When ghostty launches a command with `-e` it wont use `gtk-single-instance=true ` as described in the docs

**Comment by gptlang** [comment, marked as answer]
The initial posted issue mentioned using Hyprland.

ghostty uses `launchedFromDesktop` to determine if a shared gtk instance should be used.
Hyprland launches the process directly without using a `.desktop` file.
The solution is to set:

```ini
exec-once = ghostty --gtk-single-instance=true --quit-after-last-window-closed=false --initial-window=false

$terminal = ghostty --gtk-single-instance=true
```

in your Hyprland config.
This gets ghostty actually on par with kitty and actually faster than wezterm.

This feels good enough in the sense that going faster isn't really noticeable beyond this point, especially accounting for animations and such.

**Reply by justNeto** [reply]
This absolutely worked for me. Thank you!

**Reply by Sneethe** [reply]
Sorry but was this
>$terminal = ghostty --gtk-single-instance=true

 meant to be how ghotty is spawned? So for Hyprland that would be:
 bind = Super, Return, exec, ghostty --gtk-single-instance=true

**Reply by gptlang** [reply]
Ah by default it goes like this:
```
bind = $mainMod, Q, exec, $terminal
```
What you wrote works too. I have multiple keybinds to the terminal, one of them directly spawning tmux in it and it helps not having to repeat the args

**Reply by winterec** [reply]
Thank you

**Reply by rockethm** [reply]
Hey! (I'm a bit of a noob so sorry if this doesn't make much sense)

I see that this happens to hyprland but I'm not using hyprland and am getting the slow first start. 

How should I approach looking into it? I'm on Linux Mint if it is of any help.

**Reply by pluiedev** [reply]
You need to install what is known as an [XDG Desktop Portal](https://wiki.archlinux.org/title/XDG_Desktop_Portal) for your system, and for Cinnamon (default DE on Mint) that means installing both `xdg-desktop-portal-gtk` and `xdg-desktop-portal-xapp` I think. If you have them installed and Ghostty is still opening slowly, make sure to have `gtk-single-instance=true` in your config. If it still doesn't work, please open a new post under the "Issue Triage" category so that we can have more information about your system and help you more effectively.

Also, in the future please don't comment on old threads. There are 14 people currently subscribed to the thread and each and every one of us is getting an email for every comment unless we opt out of it. See #6937 for more information.

**Comment by Mange** [comment]
Although reusing instances work for the common case, I wonder if we could make it faster for other custom cases.

Some examples:

- Running with a custom `--class` option set. (To have custom window rules, etc.)
- Opening terminals in particular directories with `--working-directory`.
- Opening terminals for a custom command with the `-e` option.

I have keybinds to open terminals in different directories, but since that triggers new startup options it will override the single-instance option and take a long time again.

Opening a new instance takes 1.3-1.5 seconds on my 32 core, 64 GiB RAM, nvme system. Anyone have any ideas on what we could do about this? Would it be hard to allow the `-e` or `--working-directory` options to still reuse a running instance?

---

I cannot comfortably reuse instances in my setup anyway since I have a lot of custom plugins that checks which CWD the focused window has, but this all leads back to the root process and doesn't show the individual window's CWD anymore.

I'd love to get the GTK init to be faster, but I understand that's far beyond the scope of this project. Not asking for that.

**Reply by rhodes-b** [reply]
this is a very old post, that was already closed you should either find a newer one or make a new post.

But to answer your questions 1.3-1.5 seconds is very out of the normal. "Slow" would be like 500ms so something about your system is odd. have you tried tip? or ensure the desktop portals are setup correctly? (tip does not do synchronous portal / dbus interaction) (even on my old laptop that is like 7 years old ghostty takes like 500ms)

> Running with a custom --class option set. (To have custom window rules, etc.)

I believe the systemd scopes allow this im not sure though (this is only a tip feature but 1.2 will have it which will be released soon)
https://tip.ghostty.org/docs/linux/systemd

> Opening terminals in particular directories with --working-directory.

This is already another discussion https://github.com/ghostty-org/ghostty/discussions/8043 (edit: this also requires some ipc)

> Opening terminals for a custom command with the -e option.

This requires some ipc mechanism like #2353 or something similar to the +new-window action which uses dbus (im not sure dbus can actually pass args the way its done)


Again I very much suggest you make new posts (since they are separate ideas) or contribute to other more recent posts than something almost a year old

**Reply by jcollie** [reply]
> > Running with a custom --class option set. (To have custom window rules, etc.)
> 
> I believe the systemd scopes allow this im not sure though (this is only a tip feature but 1.2 will have it which will be released soon) https://tip.ghostty.org/docs/linux/systemd

You _can_ customize the class but it will run as a separate instance because that's an application-level setting that can't be changed on a window-by-window basis.

**Reply by Mange** [reply]
> this is a very old post, that was already closed you should either find a newer one or make a new post.

I apologize. I'm more used to the netiquette of keeping related discussions together and not spread them out over multiple forum threads. I'll refrain from doing it again for this repository. :-)

I figured that the matter shouldn't be "closed" when only a workaround that works for just a few zoomed in cases was presented and wanted to bring up cases where the workaround wouldn't work.

I'll see if I have time to get into Zig to help contribute to some of those follow-up issues or not. Thanks!
